Src/WindowsServer/WindowsServer.Shared/Implementation/AzureHeartbeatProperties.cs

```diff
@@ -0,0 +1,95 @@
+    using System.Threading.Tasks;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    internal class AzureHeartbeatProperties
+    {
+        // internal for testing
+        internal readonly List<string> DefaultFields = new List<string>()
+        {
+            "osType",
+            "location",
+            "name",
+            "offer",
+            "platformFaultDomain",
+            "platformUpdateDomain",
+            "publisher",
+            "sku",
+            "version",
+            "vmId",
+            "vmSize",
+            "subscriptionId",
+            "resourceGroupName"
+        };
+
+        /// <summary>
+        /// Flags that will tell us whether or not Azure VM metadata has been attempted to be gathered or not, and
+        /// if we should even attempt to look for it in the first place. Static to ensure we only ever attempt this
+        /// once for any given consuming application.
+        /// </summary>
+        private static bool isAzureMetadataCheckCompleted = false;
+
+        private IAzureMetadataRequestor azureInstanceMetadataRequestor = null;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureHeartbeatProperties"/> class.
+        /// </summary>
+        public AzureHeartbeatProperties() : this(null, false)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureHeartbeatProperties"/> class.
+        /// </summary>
+        /// <param name="azureInstanceMetadataHandler">For testing: Azure metadata request handler to use when requesting data from azure specifically. If left as null, an instance of AzureMetadataRequestor is used.</param>
+        /// <param name="resetCheckCompleteFlag">For testing: set to true to reset the check that we've already acquired this data.</param>
+        internal AzureHeartbeatProperties(IAzureMetadataRequestor azureInstanceMetadataHandler = null, bool resetCheckCompleteFlag = false)
+        {
+            this.azureInstanceMetadataRequestor = azureInstanceMetadataHandler;
+            if (this.azureInstanceMetadataRequestor == null)
+            {
+                this.azureInstanceMetadataRequestor = new AzureMetadataRequestor();
+            }
+
+            if (resetCheckCompleteFlag)
+            {
+                AzureHeartbeatProperties.isAzureMetadataCheckCompleted = false;
+            }
+        }
+
+        public bool IsKeyword(string keyword)
+        {
+            return this.DefaultFields.Contains(keyword, StringComparer.OrdinalIgnoreCase);
+        }
+
+        public async Task<bool> SetDefaultPayload(IEnumerable<string> disabledFields, IHeartbeatPropertyManager provider)
+        {
+            bool hasSetFields = false;
+
+            if (!AzureHeartbeatProperties.isAzureMetadataCheckCompleted)
+            {
+                // only ever attempt this once
+                AzureHeartbeatProperties.isAzureMetadataCheckCompleted = true;
```

Comment:
If SetDefaultPayload is invoked twice at the same time, the communication with IMS may still be attempted twice due to positive concurrency (both calls may pass a check above before they set `isAzureMetadataCheckCompleted` to true.

Reply:
Understood, and there is the possibility that this could happen regardless (what if someone instantiated two objects that call this?). 

It is unimportant however as the underlying heartbeat manager will just happily set the fields twice without issue, there is no problematic contention here.

Admittedly, this was a leftover concept from when I had it in the hearbeat manager, and I should have removed it. 

The `isAzureMetadataCheckCompleted` will now be used to reduce the likelihood it gets called twice in any single instance of `AzureHeartbeatProperties`.